### PR TITLE
Add checkout tests TA-20 - 22

### PR DESCRIPTION
**Pull request by a-kotova**

### CODE REVIEW

File: cypress/page-objects/checkoutPage.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import BasePage from './basePage';
+
+class CheckoutPage extends BasePage {
+  get total() {
+    return cy.contains('Total');
+  }
+
+  get cartQuantity() {
+    return cy.contains('SHOP').parent().find('div:last-child > span');
```
Comment (s34rching) [inline cypress/page-objects/checkoutPage.js:9]: nice

File: cypress/fixtures/products.json [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+      {
+        "id": 29,
+        "imageUrl": "https://i.ibb.co/qBcrsJg/white-vest.png",
+        "name": "White Blouse",
+        "price": 20
+      }
+    ]
+  }
+]
+}
```
Comment (s34rching) [inline cypress/fixtures/products.json]: where is an empty line?

File: cypress/integration/tests/checkout/checkout.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+const chance = new Chance();
+
+describe('Checkout', () => {
+  beforeEach(() => {
+    cy.wrap(ShopPage.pickRandomProduct(products)).as('randomProduct');
+  });
+
+  it('TA-20: Cart items count should be equal to number of added products', function () {
+    const quantity = chance.integer({ min: 1, max: 12 });
```
Comment (s34rching) [inline cypress/integration/tests/checkout/checkout.js]: please try to do the same with `lodash` which is bundled into Cypress

File: cypress/integration/tests/checkout/checkout.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import Chance from 'Chance';
+import CheckoutPage from '../../../page-objects/checkoutPage';
+import products from '../../../fixtures/products.json';
+import ShopPage from '../../../page-objects/shopPage';
+
+const chance = new Chance();
+
+describe('Checkout', () => {
+  beforeEach(() => {
+    cy.wrap(ShopPage.pickRandomProduct(products)).as('randomProduct');
```
Comment (s34rching) [inline cypress/integration/tests/checkout/checkout.js]: nope
please declare variable in a higher scope and then initialize it inside of `before` each.
your version works, but the usage of `this` is kinda confusing

Do not listen to your Indian guy ~_^)

File: cypress/page-objects/checkoutPage.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  }
+
+  getCartQuantity() {
+    return this.cartQuantity.then(($quantity) => {
+      return cy.wrap(Number.parseInt($quantity.text(), 10));
```
Comment (s34rching) [inline cypress/page-objects/checkoutPage.js:14]: nice

File: cypress/page-objects/shopPage.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  selectProduct(selectedCategory) {
+    const selectedItem = chance.pickone(selectedCategory.items);
+    selectedItem.linkUrl = selectedCategory.linkUrl;
+    return selectedItem;
+  }
+
+  pickRandomProduct(listOfProducts) {
+    const selectedCategory = this.selectCategory(listOfProducts);
+    return this.selectProduct(selectedCategory);
+  }
```
Comment (s34rching) [inline cypress/page-objects/shopPage.js]: for all these methods please create `utils` / `service` / use `support` provided by Cypress folder and move them there.
these support methods seem reusable and do not relate to specific page

File: cypress/page-objects/shopPage.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+  addProductToCart(pickedItem, quantity = 1) {
+    this.open(pickedItem.linkUrl);
+    cy.get(`div[name="${pickedItem.name}"]`).parent().find('button').invoke('show')
+      .contains('Add to Cart')
+      .then(($CTA) => {
```
Comment (s34rching) [inline cypress/page-objects/shopPage.js]: `$CTA`
`$` is not required here, it is just a wildcard showing that something is variable, you can remove it

File: cypress/page-objects/shopPage.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    return selectedItem;
+  }
+
+  pickRandomProduct(listOfProducts) {
+    const selectedCategory = this.selectCategory(listOfProducts);
+    return this.selectProduct(selectedCategory);
+  }
+
+  addProductToCart(pickedItem, quantity = 1) {
+    this.open(pickedItem.linkUrl);
```
Comment (s34rching) [inline cypress/page-objects/shopPage.js]: this line should NOT be here: it is completely different functionality - please do NOT let your methods do too much
so please split this to `navigate to page` and `then add products to cart` steps in your test

File: package.json [hunk 9]
```diff
@@ -8,10 +8,13 @@
     "dev": "concurrently --kill-others-on-fail \"yarn server\" \"yarn client\"",
     "start": "node server.js",
     "heroku-postbuild": "cd client && npm install && npm install --only=dev --no-shrinkwrap && npm run build",
-    "server:start": "wait-on http://localhost:3000"
+    "server:start": "wait-on http://localhost:3000",
+    "cy:run": "node_modules\\.bin\\cypress run -b chrome --headless",
+    "cy:run:sanity": "for /l %x in (1, 1, 10) do yarn cy:run"
```
Comment (s34rching) [inline package.json]: okay, so please create one command for unix systems b/c this is windows iteration

anyway <3

**Review by s34rching — CHANGES_REQUESTED** [review]
nea

File: package.json (hunk 9 above)
Comment (a-kotova) [reply package.json]: Done. Let me know if it works since I have only windows (should work).

File: cypress/page-objects/shopPage.js (hunk 7 above)
Comment (a-kotova) [reply cypress/page-objects/shopPage.js]: gotcha, done

File: cypress/fixtures/products.json (hunk 2 above)
Comment (a-kotova) [reply cypress/fixtures/products.json]: ooops! Done

File: cypress/integration/tests/checkout/checkout.js (hunk 4 above)
Comment (a-kotova) [reply cypress/integration/tests/checkout/checkout.js]: This time I was referring to Cypress documentation :))
I was first trying to do as you suggest, but I didn't realize that I should declare my variable outside of before Each and that's why it wasn't working for me.

File: cypress/page-objects/shopPage.js (hunk 8 above)
Comment (a-kotova) [reply cypress/page-objects/shopPage.js]: Sure, done.

File: cypress/integration/tests/checkout/checkout.js (hunk 3 above)
Comment (a-kotova) [reply cypress/integration/tests/checkout/checkout.js]: I'm in love with it. So easy and readable and doesn't require additional imports. TY <3

File: cypress/page-objects/shopPage.js (hunk 6 above)
Comment (a-kotova) [reply cypress/page-objects/shopPage.js]: done

**Review by s34rching — CHANGES_REQUESTED** [review]
NZCHT

PLEASE RUN YOUR CODE AT LEAST ONCE BEFORE SUBMITTING ON REVIEW <3

image

File: cypress/page-objects/shopPage.js [hunk 10]
```diff
@@ -0,0 +1,19 @@
+import BasePage from './basePage';
+
+class ShopPage extends BasePage {
+  navigateToCategory(pickedItem) {
+    this.open(pickedItem.linkUrl);
+  }
```
Comment (s34rching) [inline cypress/page-objects/shopPage.js]: remove this

File: package.json [hunk 11]
```diff
@@ -8,10 +8,14 @@
     "dev": "concurrently --kill-others-on-fail \"yarn server\" \"yarn client\"",
     "start": "node server.js",
     "heroku-postbuild": "cd client && npm install && npm install --only=dev --no-shrinkwrap && npm run build",
-    "server:start": "wait-on http://localhost:3000"
+    "server:start": "wait-on http://localhost:3000",
+    "cy:run": "node_modules\\.bin\\cypress run -b chrome --headless",
```
Comment (s34rching) [inline package.json:12]: btw it does not work in unix systems regarding path

File: cypress/integration/tests/checkout/checkout.js [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+describe('Checkout', () => {
+  let randomProduct;
+
+  beforeEach(() => {
+    randomProduct = pickRandomProduct(products);
+  });
+
+  it('TA-20: Cart items count should be equal to number of added products', () => {
+    const quantity = Cypress._.random(1, 12);
+    ShopPage.navigateToCategory(randomProduct);
```
Comment (s34rching) [inline cypress/integration/tests/checkout/checkout.js]: update it as you've seen

File: cypress/integration/tests/checkout/checkout.js [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    ShopPage.addProductToCart(randomProduct, quantity);
+    CheckoutPage.getCartQuantity().should('eq', quantity);
+  });
+
+  it('TA-21: Cart total should be correct', () => {
+    ShopPage.navigateToCategory(randomProduct);
```
Comment (s34rching) [inline cypress/integration/tests/checkout/checkout.js]: and here

File: cypress/integration/tests/checkout/checkout.js [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    ShopPage.addProductToCart(randomProduct);
+    ShopPage.open('/checkout');
+    CheckoutPage.getTotal().should('eq', randomProduct.price);
+  });
+
+  it('TA-22: User is able to increase product quantity', () => {
+    ShopPage.navigateToCategory(randomProduct);
```
Comment (s34rching) [inline cypress/integration/tests/checkout/checkout.js]: and here

**Review by s34rching — CHANGES_REQUESTED** [review]
ne

File: cypress/integration/tests/checkout/checkout.js (hunk 12 above)
Comment (a-kotova) [reply cypress/integration/tests/checkout/checkout.js]: done

File: cypress/integration/tests/checkout/checkout.js (hunk 13 above)
Comment (a-kotova) [reply cypress/integration/tests/checkout/checkout.js]: done

File: cypress/integration/tests/checkout/checkout.js (hunk 14 above)
Comment (a-kotova) [reply cypress/integration/tests/checkout/checkout.js]: done

File: cypress/page-objects/shopPage.js (hunk 10 above)
Comment (a-kotova) [reply cypress/page-objects/shopPage.js]: done

**Review by s34rching — APPROVED** [review]

**Pull request merged**